Source/WebCore/dom/ShadowRoot.cpp

```diff
@@ -466,4 +467,30 @@ Vector<RefPtr<WebAnimation>> ShadowRoot::getAnimations()
     });
 }
 
+bool ShadowRoot::hasReferenceTarget() const
+{
+    return !m_referenceTarget.isNull();
+}
+
+const AtomString& ShadowRoot::referenceTarget() const
+{
+    return m_referenceTarget;
 }
+
+void ShadowRoot::setReferenceTarget(const AtomString& referenceTarget)
+{
+    if (!document().settings().shadowRootReferenceTargetEnabled())
+        return;
+
+    m_referenceTarget = referenceTarget;
+}
+
+RefPtr<Element> ShadowRoot::referenceTargetElementOrHost() const
```

Comment:
I think it would be simpler to have this member function return nullptr when the reference target isn't set.
When this whole function could be inlined in the header file like this:
```cpp
RefPtr<Element> ShadowRoot::referenceTargetElementOrHost() const
{
    return m_referenceTarget ? getElementById(m_referenceTarget) : nullptr;
}
```